presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -390,6 +360,51 @@ public String toString()
         return stateMachine.toString();
     }
 
+    private class StageTaskListener
+            implements StateChangeListener<TaskInfo>
+    {
+        private long previousMemory;
+
+        @Override
+        public void stateChanged(TaskInfo taskInfo)
+        {
+            long currentMemory = taskInfo.getStats().getMemoryReservation().toBytes();
+            long deltaMemoryInBytes = currentMemory - previousMemory;
```

Comment:
Can it be called concurrently? This is exactly the same mechanism that is being used in existing `MemoryTrackingRemoteTaskFactory.UpdatePeakMemory` class. I though that only `updateMemoryUsage` should be thread safe. Could it be possible that also task peak memory tracking is broken? I can fix both of them by using `AtomicLong`